Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h

```diff
@@ -646,6 +647,211 @@ void orient_to_bound_a_volume(TriangleMesh& tm)
+
+  typedef dynamic_face_property_t<F_cc_id>                   Face_property_tag;
+  typedef typename boost::property_map<PolygonMesh, Face_property_tag>::type   Face_cc_map;
+  Face_cc_map f_cc_ids  = get(Face_property_tag(), pm);
+  F_cc_id nb_cc = connected_components(pm, f_cc_ids, parameters::face_index_map(fim));
+
+  std::vector<std::size_t> nb_faces_per_cc(nb_cc, 0);
+  for (face_descriptor f : faces(pm))
+    nb_faces_per_cc[ get(f_cc_ids, f) ]+=1;
+
+  std::map< std::pair<Point_3, Point_3>, std::vector<halfedge_descriptor> > border_hedges_map;
+  std::vector<halfedge_descriptor> border_hedges;
+  typedef typename boost::property_map<PolygonMesh, dynamic_halfedge_property_t<B_cc_id> >::type H_to_bcc_id;
+  H_to_bcc_id h_bcc_ids = get(dynamic_halfedge_property_t<B_cc_id>(), pm);
+  const B_cc_id DV(-1);
+  const B_cc_id FILTERED_OUT(-2);
+
+  // collect border halfedges
+  for (halfedge_descriptor h : halfedges(pm))
+    if ( is_border(h, pm) )
+    {
+      put(h_bcc_ids, h, DV);
+      border_hedges.push_back(h);
+    }
+
+  // compute the border cc id of all halfedges and mark those duplicated in their own cycle
+  B_cc_id bcc_id=0;
+  for (halfedge_descriptor h : border_hedges)
+  {
+    if (get(h_bcc_ids,h) == DV)
+    {
+      typedef std::map< std::pair<Point_3, Point_3>, halfedge_descriptor> Hmap;
+      Hmap hmap;
+      for (halfedge_descriptor hh : halfedges_around_face(h, pm))
+      {
+        std::pair< typename Hmap::iterator, bool > insert_res =
+          hmap.insert(
+            std::make_pair(
+              make_sorted_pair(get(vpm, source(hh, pm)),
+              get(vpm, target(hh,pm))), hh) );
+        if (insert_res.second)
+          put(h_bcc_ids, hh, bcc_id);
+        else
+        {
+          put(h_bcc_ids, hh, FILTERED_OUT);
+          put(h_bcc_ids, insert_res.first->second, FILTERED_OUT);
+        }
+      }
+      ++bcc_id;
+    }
+  }
+
+  // fill endpoints -> hedges
+  for (halfedge_descriptor h : border_hedges)
+  {
+    if ( get(h_bcc_ids, h) != FILTERED_OUT)
+      border_hedges_map[std::make_pair(get(vpm, source(h, pm)), get(vpm, target(h, pm)))].push_back(h);
+  }
+
+  // max nb of faces for a CC to be reversed
+  const std::size_t threshold =
+    parameters::choose_parameter( parameters::get_parameter(np, internal_np::maximum_number_of_faces), 0);
+
+  std::vector<bool> border_cycle_to_ignore(bcc_id, false);
+  std::vector<F_cc_id> cycle_f_cc_id(bcc_id);
+  std::vector< std::vector<F_cc_id> > patch_neighbors(nb_cc);
+
+  for (const auto& p : border_hedges_map)
+  {
+    const std::vector<halfedge_descriptor>& hedges = p.second;
+    switch(hedges.size())
+    {
+      case 1:
+        // isolated border hedge nothing to do
+      break;
+      case 2:
+      {
+        F_cc_id cc_id_0 = get(f_cc_ids, face(opposite(hedges[0], pm), pm)),
+                cc_id_1 = get(f_cc_ids, face(opposite(hedges[1], pm), pm));
+
+        if (cc_id_0!=cc_id_1)
```

Comment:
could start with a stitch_borders since you'll stitch in any case (?)

Reply:
I don't remember the details but I think it would break the logic of merging with priority